assets/arc-0052/bip32-ed25519.ts

```diff
@@ -0,0 +1,181 @@
+import { createHash, createHmac } from "crypto";
+import {
+    crypto_core_ed25519_add,
+  crypto_scalarmult_ed25519_base_noclamp
+} from "libsodium-wrappers-sumo";
+var BN = require("bn.js");
+
+/**
+ *
+ * Reference of BIP32-Ed25519 Hierarchical Deterministic Keys over a Non-linear Keyspace (https://acrobat.adobe.com/id/urn:aaid:sc:EU:04fe29b0-ea1a-478b-a886-9bb558a5242a)
+ *
+ * @see section V. BIP32-Ed25519: Specification;
+ * 
+ * A) Root keys
+ * 
+ * @param seed - 256 bite seed generated from BIP39 Mnemonic
+ * @returns - Extended root key (kL, kR, c) where kL is the left 32 bytes of the root key, kR is the right 32 bytes of the root key, and c is the chain code. Total 96 bytes
+ */
+export function fromSeed(seed: Buffer): Uint8Array {
+  // k = H512(seed)
+  let k: Buffer = createHash("sha512").update(seed).digest();
+  let kL: Buffer = k.subarray(0, 32);
+  let kR: Buffer = k.subarray(32, 64);
+
+  // While the third highest bit of the last byte of kL is not zero
+  while ((kL[31] & 0b00100000) !== 0) {
+    k = createHmac("sha512", kL).update(kR).digest();
+    kL = k.subarray(0, 32);
+    kR = k.subarray(32, 64);
+  }
+
+  // clamp
+  //Set the bits in kL as follows:
+  // little Endianess
+  kL[0] &= 0b11_11_10_00; // the lowest 3 bits of the first byte of kL are cleared
+  kL[31] &= 0b01_11_11_11; // the highest bit of the last byte is cleared
+  kL[31] |= 0b01_00_00_00; // the second highest bit of the last byte is set
+
+  // chain root code
+  // SHA256(0x01||k)
+  const c: Buffer = createHash("sha256").update(Buffer.concat([new Uint8Array([0x01]), seed])).digest();
+  return new Uint8Array(Buffer.concat([kL, kR, c]));
+}
+
+/**
+ * @see section V. BIP32-Ed25519: Specification;
+ * 
+ * subsections:
+ * 
+ * B) Child Keys
+ * and
+ * C) Private Child Key Derivation
+ * 
+ * @param extendedKey - extended key (kL, kR, c) where kL is the left 32 bytes of the root key the scalar (pvtKey). kR is the right 32 bytes of the root key, and c is the chain code. Total 96 bytes
+ * @param index - index of the child key
+ * @returns - (kL, kR, c) where kL is the left 32 bytes of the child key (the new scalar), kR is the right 32 bytes of the child key, and c is the chain code. Total 96 bytes
+ */
+export function deriveChildNodePrivate(
+  extendedKey: Uint8Array,
+  index: number
+): Uint8Array {
+  const kl: Buffer = Buffer.from(extendedKey.subarray(0, 32));
+  const kr: Buffer = Buffer.from(extendedKey.subarray(32, 64));
+  const cc: Uint8Array = extendedKey.subarray(64, 96);
+
+  const { z, childChainCode } = index < 0x80000000 ? derivedNonHardened(kl, cc, index) : deriveHardened(kl, kr, cc, index);
+
+  const chainCode = childChainCode.subarray(32, 64);
+  const zl = z.subarray(0, 32);
+  const zr = z.subarray(32, 64);
+
+  // left = kl + 8 * trunc28(zl)
+  // right = zr + kr
+
+  const left = new BN(kl, 16, "le").add(new BN(zl.subarray(0, 28), 16, "le").mul(new BN(8))).toArrayLike(Buffer, "le", 32);
```

Comment:
According to the paper, `if kL is divisible by the base order n, discard the child`. Where do we check if this is divisible?

Reply:
This is where i went and checked all other known implementations (i.e Nano, Cardano, bip32-ed25519) and none of them do anything obvious related to this. The paper doesn't state what you should do if you discard the child. 

Derivations are deterministic, if you were to reject a child, you wouldn't be able to derive further keys on that path. 

I've messaged the authors trying to get a clarification. Do you have a suggestion?

Reply:
I assume inside the wallet itself you'd simply have to flag that index as bad and move to the next index?